src/devices/SoftPwm/SoftwarePwmChannel.cs

```diff
@@ -150,42 +153,19 @@ private void Run()
                 if (_pinHighTime != TimeSpan.Zero)
```

Comment:
Can you add an assert so that `_pingHighTime` and `_pinLowTime` won't become zero while `_isRunning`

Reply:
But they can become zero while running.

It is perfectly valid to change the Duty Cycle to `1.0` or `0.0` in which case the pin will always be high or always be low.

Reply:
What I'm concerned about is that when both are zero/one this will use 100% of the CPU for no reason, for those two cases might make sense to special case to optimize CPU usage

Reply:
Why would it use 100% CPU? You can't have both be zero and so you will always hit one of the two cases and then delay for `period`.

So if your Frequency is 60, then the period is `16.67ms`, if you have a duty cycle of `0.5` you will:
* Go High
* Delay for `8.33ms`
* Go Low
* Delay for `8.33ms`
* Repeat

Likewise, if the Duty Cycle is `1.0` you would:
* Go High
* Delay for `16.66ms`
* See LowTime is 0, so do nothing
* Repeat

Reply:
For a Duty Cle of `0.0`, it would naturally be inversed:
* See HighTime is 0, so do nothing
* Go Low
* Delay for 16.66ms
* Repeat

Reply:
I meant just sanity check that they both don't become zeros, either way I'm fine with this as is

Reply:
Actually, it's possible for both to be zero.

Given there are two threads involved, the other thread can update the Duty Cycle between when `_pinHighTime` and `_pinLowTime` is read, and for that cycle it would end up with both being zero. It would then fix itself in the next cycle.

I don't think its worth trying to account for that given things will always normalize themselves in the cycle following the update (which is no more than 1s).

---

We could assert in `UpdatePulseWidthParameters` that `pinHighTicks` and `pinLowTicks` aren't both zero, but I think that's the closest to what you were asking that we could do?

Reply:
Seems like not worth to add a check